compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs

```diff
@@ -1478,23 +1489,72 @@ impl<'a, 'tcx> InferCtxtPrivExt<'tcx> for InferCtxt<'a, 'tcx> {
                 })
                 .collect(),
             None => all_impls
-                .filter_map(|def_id| {
-                    if self.tcx.impl_polarity(def_id) == ty::ImplPolarity::Negative {
-                        return None;
-                    }
-                    self.tcx.impl_trait_ref(def_id)
-                })
+                .filter(|&def_id| self.tcx.impl_polarity(def_id) != ty::ImplPolarity::Negative)
+                .filter_map(|def_id| self.tcx.impl_trait_ref(def_id))
                 .collect(),
         }
     }
 
     fn report_similar_impl_candidates(
         &self,
         impl_candidates: Vec<ty::TraitRef<'tcx>>,
+        trait_ref: ty::PolyTraitRef<'tcx>,
         err: &mut DiagnosticBuilder<'_>,
-    ) {
+    ) -> bool {
+        let def_id = trait_ref.def_id();
         if impl_candidates.is_empty() {
-            return;
+            if self.tcx.trait_is_auto(def_id)
+                || self.tcx.lang_items().items().contains(&Some(def_id))
+                || self.tcx.get_diagnostic_name(def_id).is_some()
+            {
+                // Mentioning implementers of `Copy`, `Debug` and friends is not useful.
+                return false;
+            }
+            let mut normalized_impl_candidates: Vec<_> = self
+                .tcx
+                .all_impls(def_id)
+                // Ignore automatically derived impls and `!Trait` impls.
+                .filter(|&def_id| {
+                    self.tcx.impl_polarity(def_id) != ty::ImplPolarity::Negative
+                        || self.tcx.is_builtin_derive(def_id)
+                })
+                .filter_map(|def_id| self.tcx.impl_trait_ref(def_id))
+                // Avoid mentioning type parameters.
+                .filter(|trait_ref| !matches!(trait_ref.self_ty().kind(), ty::Param(_)))
+                .map(|trait_ref| format!("\n  {}", trait_ref.self_ty()))
+                .collect();
+            normalized_impl_candidates.sort();
```

Comment:
The logic for emitting the actual diagnostic after `sort`/`dedup` feels quite similar in this case and the diagnostic emitted below, do you think we could deduplicate a little bit here?

Maybe this?

```rust
let impl_candidates = if impl_candidates.is_empty() { 
  /* what you do now up to this line */
 } else { 
  impl_candidates 
};
```

Reply:
done